packages/oauth-client/src/OAuthClient.ts

```diff
@@ -0,0 +1,228 @@
+import got, { CancelableRequest, Response } from 'got';
+import { URL } from 'url';
+
+const validateUrl = (urlString: string) => {
+  // eslint-disable-next-line no-new
+  new URL(urlString);
+};
+
+type AccessTokenResponse = Response<{
+  access_token: string,
+  refresh_token: string,
+  endpoint: string,
+  expires_in: number
+}>;
+
+const encodeCredentials = (username: string, password: string) =>
+  Buffer.from(`${username}:${password}`).toString('base64');
+
+/**
+ * A generic authorization client
+ */
+export class OAuthClient {
+  readonly clientId: string;
+  readonly clientPassword: string;
+  readonly loginUrl: string;
+  readonly redirectUri: string;
+
+  /**
+   * @param {Object} params
+   * @param {string} params.clientId - see example
+   * @param {string} params.clientPassword - see example
+   * @param {string} params.loginUrl - see example
+   * @param {string} params.redirectUri - see example
+   *
+   * @example
+   *
+   * const oAuth2Provider = new OAuthClient({
+   *   clientId: 'my-client-id',
+   *   clientPassword: 'my-client-password',
+   *   loginUrl: 'https://earthdata.login.nasa.gov',
+   *   redirectUri: 'http://my-api.com'
+   * });
+   */
+  constructor(
+    params: {
+      clientId: string,
+      clientPassword: string,
+      loginUrl: string,
+      redirectUri: string
+    }
+  ) {
+    const {
+      clientId,
+      clientPassword,
+      loginUrl,
+      redirectUri,
+    } = params;
+
+    if (!clientId) throw new TypeError('clientId is required');
+    this.clientId = clientId;
+
+    if (!clientPassword) throw new TypeError('clientPassword is required');
+    this.clientPassword = clientPassword;
+
+    if (!loginUrl) throw new TypeError('loginUrl is required');
+    validateUrl(loginUrl);
+    this.loginUrl = loginUrl;
+
+    if (!redirectUri) throw new TypeError('redirectUri is required');
+    validateUrl(redirectUri);
+    this.redirectUri = redirectUri;
```

Comment:
We could remove this validation. The child class should validate it's params before calling `super`. Didn't seem to hurt to leave it here but open to removing.

Reply:
I think we only need to validate in parent class.